### ISSUE TITLE
On Debian 12, Node only listens on IPV6

### ISSUE THREAD

**Issue by mwdiers** [saltcorn/saltcorn#2792]
On a clean base Debian 12 server install, Node binds to IPV6 but not to IPV4, and there is no Saltcorn config parameter to force the Node Server module to listen on a specific IP address. Thus Saltcorn is only listening on port 80 on IPV6 as confirmed with `lsof -i -P`:

```
saltcorn@saltcorn:~/.config$ sudo lsof -i -P
COMMAND   PID     USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
dhclient  482     root    7u  IPv4  16536      0t0  UDP *:68
avahi-dae 497    avahi   12u  IPv4  17492      0t0  UDP *:5353
avahi-dae 497    avahi   13u  IPv6  17493      0t0  UDP *:5353
avahi-dae 497    avahi   14u  IPv4  17494      0t0  UDP *:52109
avahi-dae 497    avahi   15u  IPv6  17495      0t0  UDP *:60563
node      503 saltcorn   23u  IPv6  15647      0t0  TCP *:80 (LISTEN)
sshd      520     root    3u  IPv4  15470      0t0  TCP *:22 (LISTEN)
sshd      520     root    4u  IPv6  15482      0t0  TCP *:22 (LISTEN)
postgres  532 postgres    5u  IPv4  16591      0t0  TCP localhost:5432 (LISTEN)
```

**Comment by Kilill** [comment]
Thus happens on ubuntu as well.

there are 2 causes for this 

1. server portion of uses the node http.server module and only specifies the **port** and not the **host** and according to `https://nodejs.org/api/net.html#serverlisten`:

**Comment by koobs** [comment]
Tested saltcorn installation on freebsd from scratch yesterday, and ended up stuck with saltcorn listening on localhost. Couldn't get it listening on `*:<port>` or `<specific ip>:<port>`, given the lack of a host/bind/listen paramater to the `saltcorn serve` command. 

Unlike OP, I'm stuck only able to get it to listen on localhost, which is a blocker to serving it properly. The deployment setup will be `haproxy:80|443 > backend-saltcorn:<some-unprivileged-port>`.

This is a standard deployment scenario, so saltcorn currently detecting port 80/443 in code for special casing is a hindrance. Bundle this with a current requirement to either run as root (not advised) to bind to port 80 in default setups, or workaround it by allowing unprivileged processes to bind to privileged ports via OS config, and needing to document that, adds additional unnecessary complexity and barrieres to playing with and deploying saltcorn.

A bind/listen parameter to accept a `host` to `saltcorn serve` would be greatly appreciated.

exposing the following `server.listen(options)` would additionally be amazing:

- backlog 
- ipv6Only
- reuseport

**Comment by glutamate** [comment]
@Kilill @koobs @mwdiers  so we are just looking to expose these options: https://nodejs.org/api/net.html#serverlistenoptions-callback ? 

And i assume you are not using Let'sEncrypt here ?

If so this sounds easy, i just have to insert some fields from `db.connectObj` (which is really just the config file) into the calls to server.listen in this file https://github.com/saltcorn/saltcorn/blob/master/packages/server/serve.js

**Comment by koobs** [comment]
@glutamate Proof of concept for adding host support is here https://github.com/koobs/saltcorn/commit/268a0352826eeec714a9baa3a586258016046cc8

I dared not touch glx, or get deep into the weeds trying to factor out http (80) and https(443) hardcodings for the other cases, so they remain.

Ultimately, one should be able to provide host and [http-port[,https-port]] parameters separately, regardless of the implementation (glx, non-glx, etc).

**Comment by koobs** [comment]
> [@Kilill](https://github.com/Kilill) [@koobs](https://github.com/koobs) [@mwdiers](https://github.com/mwdiers) so we are just looking to expose these options: https://nodejs.org/api/net.html#serverlistenoptions-callback ?
> 
> And i assume you are not using Let'sEncrypt here ?
> 
> If so this sounds easy, i just have to insert some fields from `db.connectObj` (which is really just the config file) into the calls to server.listen in this file https://github.com/saltcorn/saltcorn/blob/master/packages/server/serve.js

correct, not using letsencrypt, plain old backend server processes, fronted by something else. that something else will terminate ssl among other things, so we'd send through/redirect the ACME challenges ourselves

exposing as many existing stdlib options as possible will be a huge win for flexible prod deployments.

**Comment by glutamate** [comment]
#3149 - mostly borrowed your code @koobs

**Comment by koobs** [comment]
> #3149 - mostly borrowed your code @koobs 

Thanks mate, was just a PoC, amazed it wasn't rewritten wholesale 😍
